### PR TITLE
Faq code subsections

### LINKED ISSUE #48: Have subsections for FAQ/Code Questions
I think it would be a good idea to do this so users could see all the different common code answers in one look and easily click on them in the menu. It might be a good idea for other sections as well.

What do you think?

### PR DESCRIPTION
**Pull request by astrale-sharp**
In my opinion, code questions are extremely important and with that work, they are all visible in the index while reading the book. What do you think?

### CODE REVIEW

**Comment by astrale-sharp** [comment]
#48

**Review by jacobsky — CHANGES_REQUESTED** [review]
Thanks for the PR! I think I understand what you mean when you were mentioning this idea in #48

I think that there is merit to your proposal, but don't think we should split all of the code questions into separate pages. The goal is to keep the related questions grouped together as we add more and more. At some point it might make sense to break the Code Questions FAQ into separate files like "Errors", "Conversions", "GDScript Equivalents", etc. But I don't think we have quite enough to warrant this.

One feature that you might not know is that currently each of the questions can be referenced by using a `#{question-id}` which can be used to link them.

For example: 
- https://godot-rust.github.io/book/faq/code.html#why-is-there-so-much-unsafe-in-godot-rust - https://godot-rust.github.io/book/faq/code.html#what-is-the-borrowfailed-error-and-why-do-i-keep-getting-it

I think that having a table of contents would definitely be useful in each FAQ page, but putting it all in the sidebar is a little messy. I checked out the PR and it clutters up the side bar and takes away from the sections.

Instead of putting it all in the sidebar, how about adding a list of all the questions as links at the top of each page. You can link to questions in code.md similar to the following:

```md
- [What is the `BorrowFailed` error and why do I keep getting it?](#what-is-the-borrowfailed-error-and-why-do-i-keep-getting-it)
- [Why are methods on all Godot API classes using `&self` even when they should cause mutations?](#why-are-methods-on-all-godot-api-classes-using-self-even-when-they-should-cause-mutations)
- [Why is there so much `unsafe` in godot-rust?](#why-is-there-so-much-unsafe-in-godot-rust)
```

I think this would give us the best of both worlds as questions would be easily navigable in the Code Section while not cluttering up the sidebar.

**Comment by Bromeon** [comment]
@astrale-sharp While I appreciate the enthusiasm, this is now your _**7th**_ issue/PR within a couple of days.

While this pull request is on the right track, it also created an new PR instead of using https://github.com/godot-rust/book/pull/49 as requested. This increases a fair bit of management on our side, as well as confusion for other contributors, and makes progress generally harder to follow. 

For the time being, **please don't create any new issues or PRs** until the existing ones are merged.

If you have ideas, please use one of your _existing_ issues or Discord to discuss them.
Thanks a lot for your understanding!

**Comment by astrale-sharp** [comment]
Okay! Sorry again 😖

**Comment by astrale-sharp** [comment]
Indeed I didn't know about this features and I think it makes a lot of sense having a table of content instead!

File: src/faq/code.md [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+- [How can I implement static methods in GDNative?](#how-can-i-implement-static-methods-in-gdnative)
+- [How to I convert from a `Variant` or other Godot Type to the underlying Rust type?](#how-to-i-convert-from-a-variant-or-other-godot-type-to-the-underlying-rust-type)
+- [Is it possible to set subproperties of a Godot type, such as a Material?](#is-it-possible-to-set-subproperties-of-a-godot-type-such-as-a-material)
+- [What is the godot-rust equivalent of `preload`?](#what-is-the-godot-rust-equivalent-of-preload)
+- [How do I keep a reference of `Node`?](#how-do-i-keep-a-reference-of-node)
+- [What is the Rust equivalent to `onready var` in GDScript](#what-is-the-rust-equivalent-to-onready-var-in-gdscript)
+- [What types are supported for passing through the GDNative API?](#what-types-are-supported-for-passing-through-the-gdnative-api)
+- [How can I profile my code to determine the performance?](#how-can-i-profile-my-code-to-determine-the-performance)
 ## What is the `BorrowFailed` error and why do I keep getting it?
```
Comment (jacobsky) [inline src/faq/code.md:6]: Minor nitpick: Could you add add a space above the header for consistency?

**Review by jacobsky — CHANGES_REQUESTED** [review]
Just one minor change and then we could be good to go. Optionally, if you have time, would it also be possible to do the same for the other parts of the FAQ? I believe that it would be quite useful for all the sections in addition toi the code subsection.

File: src/faq/code.md (hunk 1 above)
Comment (astrale-sharp) [reply src/faq/code.md:6]: Done

**Comment by astrale-sharp** [comment]
Sure thing!

**Comment by astrale-sharp** [comment]
I'm not super github-experienced, I just have to "request for review" right?

**Review by jacobsky — APPROVED** [review]
Thank you very much for the additions.
@Bromeon I'm satisfied with the changes. If you have any concerns, please let me know otherwise I'll go ahead and merge the PR.

**Comment by astrale-sharp** [comment]
Wonderful! Happy to help!

File: src/faq/code.md [hunk 2]
```diff
@@ -1,5 +1,20 @@
 # FAQ: Common code questions
 
+## Table of contents 
+
+- [What is the `BorrowFailed` error and why do I keep getting it?](#what-is-the-borrowfailed-error-and-why-do-i-keep-getting-it)
+- [Why are methods on all Godot API classes using `&self` even when they should cause mutations?](#why-are-methods-on-all-godot-api-classes-using-self-even-when-they-should-cause-mutations)
+- [Why is there so much `unsafe` in godot-rust?](#why-is-there-so-much-unsafe-in-godot-rust)
+- [As a native script type needs to implement `fn new(owner: &Node) -> Self`, is it possible to pass additional arguments to `new`?](#as-a-native-script-type-needs-to-implement-fn-new(owner:-&node)-->-self-is-it-possible-to-pass-additional-arguments-to-new)
```
Comment (Bromeon) [inline src/faq/code.md]: This link is broken.
Should be:
```
#as-a-native-script-type-needs-to-implement-fn-newowner-node---self-is-it-possible-to-pass-additional-arguments-to-new
```

**Review by Bromeon — CHANGES_REQUESTED** [review]
Thanks a lot for the additions! 🙂 
There's one broken link. Apart from that, this PR seems good for now.

It also shows that doing this manually is extremely brittle, even changing minor wording in a subtitle will break the links. The process also takes a lot of effort; one has to:
* click the section's title
* copy the URL part past the `#` and paste it to the ToC
* copy the title's source code (not rendered) and paste it to the ToC

and think of all that whenever something changes, is reordered, added or removed.

In the future, we should look into automating this. [**mdbook-toc**](https://github.com/badboy/mdbook-toc) is the most promising solution I found in a quick search. This topic has been brought up [countless times in `mdbook` itself](https://github.com/rust-lang/mdBook/issues/1523#issuecomment-841004963) and ignored for over half a decade, so I think it's fair to say it would need a third-party solution or something we build ourselves.

File: src/faq/code.md (hunk 2 above)
Comment (astrale-sharp) [reply src/faq/code.md]: That one eluded me, thank you!

**Comment by astrale-sharp** [comment]
Okay, the link is now fixed but I would be interested to look into mdbook-toc so the code is more easily maintained. Should we wait and should i force push here when i get something working?

**Comment by astrale-sharp** [comment]
Okay, this was extremly easy to set up, as you can see in my local branch [here](https://github.com/astrale-sharp/book/tree/faq-code-subsections-mdbook-toc)

Please note that the depth level is defined in book.toml, in section [preprocessor.toc] by the field max-level and is currently set to 2 so it behaves the same as now.

File: src/faq/code.md [hunk 3]
```diff
@@ -1,5 +1,20 @@
 # FAQ: Common code questions
 
+## Table of contents 
+
+- [What is the `BorrowFailed` error and why do I keep getting it?](#what-is-the-borrowfailed-error-and-why-do-i-keep-getting-it)
+- [Why are methods on all Godot API classes using `&self` even when they should cause mutations?](#why-are-methods-on-all-godot-api-classes-using-self-even-when-they-should-cause-mutations)
+- [Why is there so much `unsafe` in godot-rust?](#why-is-there-so-much-unsafe-in-godot-rust)
+- [As a native script type needs to implement `fn new(owner: &Node) -> Self`, is it possible to pass additional arguments to `new`?](#as-a-native-script-type-needs-to-implement-fn-newowner-node---self-is-it-possible-to-pass-additional-arguments-to-new)
+- [How can I implement static methods in GDNative?](#how-can-i-implement-static-methods-in-gdnative)
+- [How to I convert from a `Variant` or other Godot Type to the underlying Rust type?](#how-to-i-convert-from-a-variant-or-other-godot-type-to-the-underlying-rust-type)
```
Comment (bluenote10) [inline src/faq/code.md]: Same fix here: _How do I ..._

**Comment by jacobsky** [comment]
> 
> 
> Okay, this was extremly easy to set up, as you can see in my local branch [here](https://github.com/astrale-sharp/book/tree/faq-code-subsections-mdbook-toc)
> 
> Please note that the depth level is defined in book.toml, in section [preprocessor.toc] by the field max-level and is currently set to 2 so it behaves the same as now.

That looks like an excellent change, apologies for making you undo all the manual work. In that case, we should use the mdbook-toc so that the links are automatically preprocessed. That will save a lot of work in the future, if you want to add that to this and modify the table of contents, I'd be happy to test locally and then merge the changes.

**Comment by astrale-sharp** [comment]
No worries!
I'm struggling with git and github to push my new branch here... it closed the PR automatically and i dont know how to reopen it... help?
@jacobsky

**Comment by Bromeon** [comment]
Sorry, didn't mean to create extra work -- I thought we could leave the manual ToC until we get to an automated solution. But if you already got something working, even better 👍🏼 

Regarding PR, this is an annoying GitHub feature that it auto-closes PRs when you remove commits. I can't reopen it either, unfortunately. Maybe you can revert the deletion with this button? 
grafik

If that doesn't work, I found a manual how to restore the branch:
https://garrytrinder.github.io/2020/05/recover-pr-after-accidentally-deleting-source-branch

**Comment by astrale-sharp** [comment]
I cant restore the branch since it was deleted and recreated again (so it already exist) but thanks to your link i can reopen my PR!
EDIT : maybe not, still struggling ^^'

**Comment by astrale-sharp** [comment]
I have vanquished git! This should be good to go

**Comment by Bromeon** [comment]
Awesome, just tried it out locally, seems to work great! 😀

One last request: since users now need to install `mdbook-toc`, could you update the ReadMe.md from
```md
$ cargo install mdbook
$ mdbook build
```
to
```
$ cargo install mdbook mdbook-toc
$ mdbook build
```
?

If possible, in the same commit (`git commit --amend`).
Thank you very much!

**Comment by astrale-sharp** [comment]
Great Idea! It's done!

**Comment by astrale-sharp** [comment]
Great idea! It's done!

**Comment by Bromeon** [comment]
Looks good to me!
@jacobsky Anything to add? Otherwise you can merge 🚀

**Review by Bromeon — APPROVED** [review]

**Comment by astrale-sharp** [comment]
Thanks much ❤️
we can merge as far as I'm concerned!

**Comment by jacobsky** [comment]
Looks good to me! @astrale-sharp thank you for the help! I'll merge it now.

**Pull request merged**